libgadget/cooling_qso_lightup.c

```diff
@@ -0,0 +1,642 @@
+    /* If the desired ionization fraction is above a threshold (by default 0.95)
+     * ionize all particles*/
+    if(desired_ion_frac > QSOLightupParams.heIIIreion_finish_frac) {
+        int i, nionized=0;
+        int64_t nion_tot=0;
+        #pragma omp parallel for reduction(+: nionized)
+        for (i = 0; i < PartManager->NumPart; i++){
+            if (P[i].Type == 0)
+                nionized += ionize_single_particle(i);
+        }
+        sumup_large_ints(1, &nionized, &nion_tot);
+        message(0, "HeII: Helium ionization finished, flash-ionizing %ld particles (%g of total)\n", nion_tot, (double) nion_tot /(double) n_gas_tot);
+    }
+
+    double rhobar = All.CP.OmegaBaryon * (3 * HUBBLE * All.CP.HubbleParam * HUBBLE * All.CP.HubbleParam)/ (8 * M_PI * GRAVITY) / pow(1 + redshift,3);
+    double totbubblegasmass = 4 * M_PI / 3. * pow(QSOLightupParams.mean_bubble, 3) * rhobar;
+    /* Total expected ionizations if the bubbles do not overlap at all
+     * and the bubble is at mean density.*/
+    int64_t non_overlapping_bubble_number = n_gas_tot * totbubblegasmass / All.CP.OmegaBaryon;
+    double initionfrac = gas_ionization_fraction();
+    double curionfrac = initionfrac;
+
+    if(curionfrac < desired_ion_frac) {
+        ncand = build_qso_candidate_list(&qso_cand, fof);
+        walltime_measure("/HeIII/Find");
+    }
+
+    int64_t ncand_before = count_QSO_halos(ncand, &ncand_tot, MPI_COMM_WORLD);
+    int iteration;
+
+    for(iteration = 0; curionfrac < desired_ion_frac; iteration++){
+        /* Get a new quasar*/
+        int new_qso = choose_QSO_halo(ncand, &ncand_before, &ncand_tot, fof->TotNgroups+iteration);
+        if(new_qso >= ncand)
+            endrun(12, "HeII: QSO %d > no. candidates %d! Cannot happen\n", new_qso, ncand);
+        /* Make sure someone has a quasar*/
+        if(ncand_tot == 0) {
+            if(desired_ion_frac - curionfrac > 0.1)
+                message(0, "HeII: Ionization fraction %g less than desired ionization fraction of %g because not enough quasars\n", curionfrac, desired_ion_frac);
+            break;
+        }
+        /* Do the ionizations with a tree walk*/
+        int64_t n_ionized = ionize_all_part(new_qso, qso_cand, fof, tree);
+        int64_t tot_qso_ionized = 0;
+        /* Check that the ionization fraction changed*/
+        MPI_Allreduce(&n_ionized, &tot_qso_ionized, 1, MPI_INT64, MPI_SUM, MPI_COMM_WORLD);
+        curionfrac += (double) tot_qso_ionized / (double) n_gas_tot;
+        tot_n_ionized += tot_qso_ionized;
+        if(new_qso > 0)
+            message(1, "HeII: Quasar %d changed the HeIII ionization fraction to %g, ionizing %ld\n", qso_cand[new_qso], curionfrac, tot_qso_ionized);
+        /* Break the loop if we do not ionize enough particles this round.
+         * Try again next timestep when we will hopefully have new BHs.*/
+        if(tot_qso_ionized < 0.01 * non_overlapping_bubble_number && iteration > 10)
```

Comment:
might be cleaner if iteration > 10 is moved to 564. Then we see the loop is bound by 10 iteration without parsing the loop body.

Reply:
oops. Except if the ionization has been efficient then we'll keep doing it forever -- is this behavior intended?

Reply:
The idea is that we exit the loop if ionization is getting slow and if we have already tried 10 times. This is to guard against picking an already ionized quasar once or twice, and no ionization happening. If the ionization is efficient then the loop will exit only when we have ionized enough, which is intentional. The loop will also eventually exit when we run out of quasars.